enhancements/20241211-upgrade-enhancement.md

```diff
@@ -0,0 +1,997 @@
+	UpgradePlanProgressing string = "Progressing"
+	// Degraded means the progress of the upgrade is stalled due to issues.
+	UpgradePlanDegraded string = "Degraded"
+)
+
+const (
+	NodeStateImagePreloading     string = "ImagePreloading"
+	NodeStateImagePreloaded      string = "ImagePreloaded"
+	NodeStateKubernetesUpgrading string = "KubernetesUpgrading"
+	NodeStateKubernetesUpgraded  string = "KubernetesUpgraded"
+	NodeStateOSUpgrading         string = "OSUpgrading"
+	NodeStateOSUpgraded          string = "OSUpgraded"
+)
+
+const (
+	UpgradePlanPhaseInitializing       UpgradePlanPhase = "Initializing"
+	UpgradePlanPhaseInitialized        UpgradePlanPhase = "Initialized"
+	UpgradePlanPhaseISODownloading     UpgradePlanPhase = "ISODownloading"
+	UpgradePlanPhaseISODownloaded      UpgradePlanPhase = "ISODownloaded"
+	UpgradePlanPhaseRepoCreating       UpgradePlanPhase = "RepoCreating"
+	UpgradePlanPhaseRepoCreated        UpgradePlanPhase = "RepoCreated"
+	UpgradePlanPhaseMetadataPopulating UpgradePlanPhase = "MetadataPopulating"
+	UpgradePlanPhaseMetadataPopulated  UpgradePlanPhase = "MetadataPopulated"
+	UpgradePlanPhaseImagePreloading    UpgradePlanPhase = "ImagePreloading"
+	UpgradePlanPhaseImagePreloaded     UpgradePlanPhase = "ImagePreloaded"
+	UpgradePlanPhaseClusterUpgrading   UpgradePlanPhase = "ClusterUpgrading"
+	UpgradePlanPhaseClusterUpgraded    UpgradePlanPhase = "ClusterUpgraded"
+	UpgradePlanPhaseNodeUpgrading      UpgradePlanPhase = "NodeUpgrading"
+	UpgradePlanPhaseNodeUpgraded       UpgradePlanPhase = "NodeUpgraded"
+	UpgradePlanPhaseCleaningUp         UpgradePlanPhase = "CleaningUp"
+	UpgradePlanPhaseCleanedUp          UpgradePlanPhase = "CleanedUp"
+
+	UpgradePlanPhaseSucceeded UpgradePlanPhase = "Succeeded"
+	UpgradePlanPhaseFailed    UpgradePlanPhase = "Failed"
+)
+
+type NodeUpgradeStatus struct {
+	State   string `json:"state,omitempty"`
+	Reason  string `json:"reason,omitempty"`
+	Message string `json:"message,omitempty"`
+}
+
+// UpgradePlanPhase defines what overall phase UpgradePlan is in
+type UpgradePlanPhase string
+
+type UpgradePlanPhaseTransitionTimestamp struct {
+	Phase                    UpgradePlanPhase `json:"phase"`
+	PhaseTransitionTimestamp metav1.Time      `json:"phaseTransitionTimestamp"`
+}
+
+type ReleaseMetadata struct {
+	Harvester            string `json:"harvester,omitempty"`
+	HarvesterChart       string `json:"harvesterChart,omitempty"`
+	OS                   string `json:"os,omitempty"`
+	Kubernetes           string `json:"kubernetes,omitempty"`
+	Rancher              string `json:"rancher,omitempty"`
+	MonitoringChart      string `json:"monitoringChart,omitempty"`
+	MinUpgradableVersion string `json:"minUpgradableVersion,omitempty"`
+}
+
+// UpgradePlanSpec defines the desired state of UpgradePlan
+type UpgradePlanSpec struct {
+	// version refers to the corresponding version resource in the same namespace.
+	// +required
+	Version string `json:"version"`
```

Comment:
Maybe consider a `Version` type (instead of just `string`) to allow for versions comparison and syntax validation. See e.g., github.com/hashicorp/go-version. Similarly, for `ReleaseMetadata.[Harvester|Kubernetes|Rancher]`.